ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,943 @@ PHP_FUNCTION(bcscale)
+	bcmath_number_obj_handlers.unset_property = bcmath_number_unset_property;
+	bcmath_number_obj_handlers.has_property = bcmath_number_has_property;
+	bcmath_number_obj_handlers.read_property = bcmath_number_read_property;
+	bcmath_number_obj_handlers.get_properties_for = bcmath_number_get_properties_for;
+	bcmath_number_obj_handlers.cast_object = bcmath_number_cast_object;
+}
+
+static zend_always_inline void bcmath_number_add_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	*ret = bc_add(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline void bcmath_number_sub_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	*ret = bc_sub(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline zend_result bcmath_number_mul_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = n1_full_scale + n2_full_scale;
+		if (UNEXPECTED(CHECK_RET_SCALE_OVERFLOW(*scale, n1_full_scale))) {
+			zend_value_error("scale of the result is too large");
+			return FAILURE;
+		}
+	}
+	*ret = bc_multiply(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_div_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = n1_full_scale + BC_MATH_NUMBER_EXPAND_SCALE;
+		if (UNEXPECTED(CHECK_RET_SCALE_OVERFLOW(*scale, n1_full_scale))) {
+			zend_value_error("scale of the result is too large");
+			return FAILURE;
+		}
+	}
+	if (!bc_divide(n1, n2, ret, *scale)) {
+		zend_throw_exception_ex(zend_ce_division_by_zero_error, 0, "Division by zero");
+		return FAILURE;
+	}
+	bc_rm_trailing_zeros(*ret);
+	if (auto_scale) {
+		size_t diff = *scale - (*ret)->n_scale;
+		*scale -= diff > BC_MATH_NUMBER_EXPAND_SCALE ? BC_MATH_NUMBER_EXPAND_SCALE : diff;
+	}
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_mod_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	if (!bc_modulo(n1, n2, ret, *scale)) {
+		zend_throw_exception_ex(zend_ce_division_by_zero_error, 0, "Modulo by zero");
+		return FAILURE;
+	}
+	bc_rm_trailing_zeros(*ret);
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_pow_internal(
```

Comment:
I'm not too keen on sprinkling zend_always_inline everywhere, especially on large functions.